### PR TITLE
Clean up wrapper for a few training type plugins

### PR DESCRIPTION
**Pull request by four4fish**
## What does this PR do?
1. Remove the direct reference to [DistributedDataParallel](https://pytorch.org/docs/master/generated/torch.nn.parallel.DistributedDataParallel.html) from ddp and ddp_spawn, added wrapper in Parallel plugin (abstract training type that user can extend)
2. Avoid rewrapping the lightningModules in ddp, ddp_spawn, sharded, and sharded_spawn

Tests: python -m pytest -v tests/plugins/ 
75 passed, 77 skipped, 62 warnings on gpu machine

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/parallel.py [hunk 1]
```diff
@@ -134,6 +134,9 @@ def block_backward_sync(self):
         else:
             yield None
 
+    def _config_model(self, model: "pl.LightningModule", device_ids: str, **ddp_kwarg) -> None:
+        self._model = DistributedDataParallel(model, device_ids=device_ids, **ddp_kwarg)
+
     def teardown(self) -> None:
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/parallel.py:140]: we shouldn't introduce DDP into the parallel plugin - rather we should do the opposite, and ensure parallel plugin stays as generic as possible. DDP is only one implementation, but other use cases may extend implement strategies using the parallel plugin very differently

**Review by ananthsub — CHANGES_REQUESTED** [review]

File: pytorch_lightning/plugins/training_type/parallel.py (hunk 1 above)
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/parallel.py:140]: Got you! That make sense, will update to the other way

File: pytorch_lightning/plugins/training_type/parallel.py [hunk 2]
```diff
@@ -137,9 +137,6 @@ def block_backward_sync(self):
...
         # Un-reference the wrapper if any was used.
         # todo (tchaton): Add support for all plugins.
-        if isinstance(self.model, DistributedDataParallel):
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/parallel.py:140]: We can also delete the comment above this

File: pytorch_lightning/plugins/training_type/parallel.py [hunk 3]
```diff
@@ -135,11 +135,6 @@ def block_backward_sync(self):
             yield None
 
     def teardown(self) -> None:
-        # Un-reference the wrapper if any was used.
-        # todo (tchaton): Add support for all plugins.
-        if isinstance(self.model, DistributedDataParallel):
-            self.model = self.lightning_module
-
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/parallel.py:142]: @tchaton from https://github.com/PyTorchLightning/pytorch-lightning/pull/8490/ - do you recall why unwrapping this saves memory in the teardown?

doing this with DDP is at odds with https://github.com/PyTorchLightning/pytorch-lightning/issues/8593 

if there are memory leaks intrinsic in the approach of 8593, should we reconsider this? @SeanNaren

Comment (tchaton) [reply pytorch_lightning/plugins/training_type/parallel.py:142]: Hey @ananthsub. It seems the DDP Reducer was holding some GPU memory and had to be deleted to get back to the initial state.

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 4]
```diff
@@ -388,3 +388,13 @@ def register_plugins(cls, plugin_registry: Dict) -> None:
...
+
+    def teardown(self) -> None:
+        if isinstance(self.model, DistributedDataParallel):
+            self.model = self.lightning_module
+
+        if self.on_gpu:
+            # GPU teardown
+            self.lightning_module.cpu()
+            # clean up memory
+            torch.cuda.empty_cache()
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:400]: Can we keep this part in the parallel plugin and call the super classes teardown here?
This is not specific to DDP but to every module basically.

**Review by justusschock — APPROVED** [review]

File: tests/plugins/test_ddp_spawn_plugin.py [hunk 5]
```diff
@@ -77,3 +78,22 @@ def test_ddp_spawn_extra_parameters(tmpdir):
     trainer.fit(model, datamodule=dm)
     assert trainer.callback_metrics[val_name] == torch.tensor(val)
     assert model.test_val == "test_val"
+
+
+class BoringModelDDP(BoringModel):
+    def on_train_start(self) -> None:
+        """Check model wrapped by DistributedDataParallel at configurate_ddp."""
+        assert isinstance(self.trainer.model, DistributedDataParallel)
```
Comment (awaelchli) [inline tests/plugins/test_ddp_spawn_plugin.py:86]: we already have tests that check that

File: tests/plugins/test_ddp_spawn_plugin.py [hunk 6]
```diff
@@ -77,3 +78,22 @@ def test_ddp_spawn_extra_parameters(tmpdir):
...
+
+    def on_fit_end(self) -> None:
+        """Check teardown function at the end of training."""
+        assert self.trainer.model.device == torch.device("cpu")
+        assert isinstance(self.trainer.model, LightningModule)
```
Comment (awaelchli) [inline tests/plugins/test_ddp_spawn_plugin.py:91]: don't we want these asserts in the main test below after fit?

File: tests/plugins/test_ddp_spawn_plugin.py [hunk 7]
```diff
@@ -77,3 +78,22 @@ def test_ddp_spawn_extra_parameters(tmpdir):
...
+
+
+@RunIf(min_gpus=1)
```
Comment (awaelchli) [inline tests/plugins/test_ddp_spawn_plugin.py:94]: was the intention to test on gpu or cpu?
if no, remove this RunIf decorator, if yes, adjust the trainer arguments below.

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:400]: +1

Comment (four4fish) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:400]: Yeah, the 396-400 still exist in parallel plugin. Added override teardown function in ddp plugins to unwrap the DistributedDataParallel model and get the lightning module first. I read through the comments history (@ananthsub @tchaton), seems that without unwrap there are gpu memory to held in ddp reducer? 
What's your recommendation? Will it be better to move the unwrap logic into other ddp functions and keep the teardown logic only in the super class (parallel)?

File: tests/plugins/test_ddp_spawn_plugin.py (hunk 6 above)
Comment (four4fish) [reply tests/plugins/test_ddp_spawn_plugin.py:91]: Yeah, that will work too! Could you share some insight about which way is better? I think they are the same logic, is that having test in main test the preferred way? thanks~~

File: tests/plugins/test_ddp_spawn_plugin.py (hunk 7 above)
Comment (four4fish) [reply tests/plugins/test_ddp_spawn_plugin.py:94]: Good catch! Yeah, it intended to test GPU teardown, will add the param in trainer, thank!

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 4 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:400]: @justusschock @awaelchli - i prefer keeping the parallel plugin as generic as possible as an interface. in the parent TrainingTypePlugin, `teardown` is an abstract method. 

the reason being, model parallel use cases have to extend the parallel plugin right now. for these use cases which have parts of the model on cpu and gpu, automatically moving the full lightning module to cpu at the end of training may not be desirable. keeping `teardown` as an abstract method here enforces that users should provide an implementation here

**Review by tchaton — APPROVED** [review]
LGTM ! But I believe we already have test for this.

File: tests/plugins/test_ddp_spawn_plugin.py (hunk 6 above)
Comment (awaelchli) [reply tests/plugins/test_ddp_spawn_plugin.py:91]: I think it's better to directly do the assert after the fit call because

1) nothing guarantees us here that on_fit_end actually runs. It's supposed to be a hook yes, but as far as the test is concerned here we have no way of knowing. it the hook doesn't run, the test will still pass!
2) I would not rely on the teardown/on_fit_end order to be the same in the future. especially since the recent changes where sometimes the teardown runs after on_fit_end and sometimes before. cc @ananthsub 
3) also if we do it directly in the main test we don't need this special model class just for testing

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 4 above)
Comment (justusschock) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:400]: @ananthsub re1: Sounds okay to me, but might be easier to forget it that way.

re2: Why wouldn't you want to completely move it to cpu?

File: tests/plugins/test_ddp_spawn_plugin.py (hunk 6 above)
Comment (ananthsub) [reply tests/plugins/test_ddp_spawn_plugin.py:91]: > I would not rely on the teardown/on_fit_end order to be the same in the future. especially since the recent changes where sometimes the teardown runs after on_fit_end and sometimes before

IIRC `teardown` should always happen after `on_fit_end` : https://github.com/PyTorchLightning/pytorch-lightning/blob/538e743f17c7da4624c902f762922e2837661818/pytorch_lightning/plugins/training_type/ddp_spawn.py#L208-L214

@awaelchli do you see this isn't the case somewhere? how should we enforce this as an invariant of the loop?

File: tests/plugins/test_ddp_plugin.py [hunk 8]
```diff
@@ -69,3 +69,22 @@ def test_ddp_barrier_non_consecutive_device_ids(barrier_mock, tmpdir):
...
+        assert self.trainer.model.device == torch.device("cpu")
+        assert isinstance(self.trainer.model, LightningModule)
+
+
+@RunIf(min_gpus=1)
+def test_ddp_plugin_teardown():
+    """Tests with ddp plugin."""
+    trainer = Trainer(num_processes=2, accelerator="ddp", fast_dev_run=True)
+    model = BoringModelDDP()
+    trainer.fit(model)
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply tests/plugins/test_ddp_plugin.py:90]: my recommendation is this

```python
plugin = DDPPluing(...)
model = BoringModel()
model.trainer = Mock()  # if needed
plugin.model = model

plugin.configure_ddp()
plugin.teardown()

assert model.device == cpu
assert not isinstance(plugin.model, DistributedDataParallel)
```
for a rudimentary unit test that looks at teardown.
haven't tested it

[the remark this answers is not in the document]
Comment (awaelchli) [reply tests/plugins/test_ddp_plugin.py:90]: I think, whenever possible, we should avoid using callbacks and hooks to test our code.

**Review by Borda — APPROVED** [review]

File: tests/plugins/test_ddp_plugin.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (four4fish) [reply tests/plugins/test_ddp_plugin.py:90]: Yeah, +1, agree avoid using hooks for test, targeted test without dependence is better. 
I feel mock everything before and after specific function is the best way to unit test, but seems our unit test is more on trainer function level, not for individual plugin functions. Do you guys see value that having unit test with more granularity?

**Comment by four4fish** [comment]
Remove the unit test as it's already covered. Will add more unit test if it's needed after https://github.com/PyTorchLightning/pytorch-lightning/pull/9096

File: tests/plugins/test_ddp_spawn_plugin.py (hunk 6 above)
Comment (awaelchli) [reply tests/plugins/test_ddp_spawn_plugin.py]: Yes @ananthsub for ddp spawn _call_teardown_hook runs before on_fit_end [(trainer.py line 1011)](https://github.com/PyTorchLightning/pytorch-lightning/blob/de57feff445744f431eaef29182b1bcbf58a1924/pytorch_lightning/trainer/trainer.py#L1017) and for anything not ddp_spawn it runs after. 

I can't find the PR anymore but these are repeated patterns that we have seen in the past and that will happen in the future, because PL community and PL team cannot agree if they want consistent hook call order or plugins that can do whatever they want.

Comment (ananthsub) [reply tests/plugins/test_ddp_spawn_plugin.py]: I see. this is the same question about why `on_fit_start` is not part of the loop. IMO consistency here is more critical. Let's file a separate issue/discussion to track this
